### PR TITLE
Add a constructor which takes in a builder

### PR DESCRIPTION
**Pull request by mc-lovin**
No of parameters you can pass in a function are limited to 255 in java, The proposal is to change the constructor to accept builders once the number of parameters exceed 16.

```
  public Message(List<Integer> field_1, List<Integer> field_2, Integer foo, Integer bar,
      ByteString unknownFields) {
    super(ADAPTER, unknownFields);
    if (Internal.countNonNull(foo, bar) > 1) {
      throw new IllegalArgumentException("at most one of foo, bar may be non-null");
    }
    this.field_1 = Internal.immutableCopyOf("field_1", field_1);
    this.field_2 = Internal.immutableCopyOf("field_2", field_2);
    this.foo = foo;
    this.bar = bar;
  }
```
changes to 
```
  public Message(Builder builder, ByteString unknownFields) {
    super(ADAPTER, unknownFields);
    if (Internal.countNonNull(builder.foo, builder.bar) > 1) {
      throw new IllegalArgumentException("at most one of builder.foo, builder.bar may be non-null");
    }
    this.field_1 = Internal.immutableCopyOf("field_1", builder.field_1);
    this.field_2 = Internal.immutableCopyOf("field_2", builder.field_2);
    this.foo = builder.foo;
    this.bar = builder.bar;
  }
```

### CODE REVIEW

File: wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 1]
```diff
@@ -1257,19 +1286,30 @@ private MethodSpec messageFieldsAndUnknownFieldsConstructor(
...
+      if (constructorTakesAllFields) {
+        ParameterSpec.Builder param = ParameterSpec.builder(javaType, fieldName);
+        if (emitAndroid && field.isOptional()) {
+          param.addAnnotation(NULLABLE);
+        }
+        result.addParameter(param.build());
       }
-      result.addParameter(param.build());
+
       if (field.isRepeated() || field.type().isMap()) {
```
Comment (swankjesse) [inline wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: nice

File: wire-tests/src/test/java/com/squareup/wire/java/JavaGeneratorTest.java [hunk 2]
```diff
@@ -66,6 +70,26 @@
...
+  @Test public void tooManyFieldsTest() throws Exception {
+    StringBuilder s = new StringBuilder();
+    for (int i = 1; i < 257; i++) {
+      s.append("  repeated int32 field_" + i + " = " + i + ";\n");
+    }
+    RepoBuilder repoBuilder = new RepoBuilder()
+        .add("message.proto", ""
+            + "message Message {\n"
+            + s.toString()
+            + "    oneof oneof_name {\n"
```
Comment (swankjesse) [inline wire-tests/src/test/java/com/squareup/wire/java/JavaGeneratorTest.java]: nice

File: wire-tests/MessageWithManyFields [hunk 3]
```diff
@@ -0,0 +1,6300 @@
...
+import java.io.IOException;
+import java.lang.Integer;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.lang.StringBuilder;
+import java.util.List;
+import okio.ByteString;
+
+public final class Message extends com.squareup.wire.Message<Message, Message.Builder> {
```
Comment (swankjesse) [inline wire-tests/MessageWithManyFields]: I think it’s going to be annoying to have this checked in. Drop it?

File: wire-tests/src/test/java/com/squareup/wire/java/JavaGeneratorTest.java [hunk 4]
```diff
@@ -66,6 +70,26 @@
...
+            + "       int32 foo = 257;\n"
+            + "       int32 bar = 258;\n"
+            + "    }\n"
+            + "}\n");
+    String generatedCode = repoBuilder.generateCode("Message");
+    Path file = Paths.get("MessageWithManyFields");
+    String message = new String(Files.readAllBytes(file), UTF_8);
+    assertThat(generatedCode).isEqualTo(message);
```
Comment (swankjesse) [inline wire-tests/src/test/java/com/squareup/wire/java/JavaGeneratorTest.java]: Tempted to just confirm that it takes a builder.

File: wire-tests/MessageWithManyFields (hunk 3 above)
Comment (JakeWharton) [reply wire-tests/MessageWithManyFields]: We already have a mechanism for golden files which have messages with more than 16 messages and which need updated via the script in the root of the repo. This should sufficiently test the functionality of this PR without any added unit tests.

Comment (mc-lovin) [reply wire-tests/MessageWithManyFields]: I see, yes it tells why the tests are failing, so I am going to nuke this file and change one of those files

**Comment by mc-lovin** [comment]
@JakeWharton updated, in case you want to take a final pass :)

**Pull request merged**